### PR TITLE
Ruby: PrintAST: more stable order for synthesized nodes

### PR DESCRIPTION
**Pull request by aibaars**
My experimental PR https://github.com/github/codeql/pull/9936 indicates that there is likely to be a [test failure](https://github.com/github/codeql/runs/7634387028?check_suite_focus=true#step:5:131) in the future caused by differences in AstNode ordering. 

This PR ensures that the AstDesugar tests pass with both the current and the nightly release.

@hvitved

### CODE REVIEW

File: ruby/ql/lib/codeql/ruby/printAst.qll [hunk 1]
```diff
@@ -112,13 +113,22 @@ class PrintRegularAstNode extends PrintAstNode, TPrintRegularAstNode {
     )
   }
 
+  private int getSynthAstNodeIndex() {
+    not astNode.isSynthesized() and result = -10
+    or
+    astNode = getSynthChild(astNode.getParent(), result)
+  }
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/printAst.qll:120]: Alternatively, to get rid of the magic constant:

```suggestion
  private int nonSynthIndex() { result = min([-1, any(int i | exists(getSynthChild(_, i)))]) - 1 }

  private int getSynthAstNodeIndex() {
    not astNode.isSynthesized() and result = nonSynthIndex()
    or
    astNode = getSynthChild(astNode.getParent(), result)
  }
```

**Review by hvitved — COMMENTED** [review]
One minor comment, otherwise LGTM.

**Review by hvitved — APPROVED** [review]

File: ruby/ql/lib/codeql/ruby/printAst.qll [hunk 2]
```diff
@@ -112,13 +113,24 @@
     )
   }
 
+  private int nonSynthIndex() { result = min([-1, any(int i | exists(getSynthChild(_, i)))]) - 1 }
+
+  private int getSynthAstNodeIndex() {
+    not astNode.isSynthesized() and result = nonSynthIndex()
```
[the remark this answers is not in the document]
Comment (hvitved) [reply ruby/ql/lib/codeql/ruby/printAst.qll:119]: Whoops, we need to move `nonSynthIndex` out of the class.

**Review by hvitved — CHANGES_REQUESTED** [review]
`nonSynthIndex` needs to be a top-level predicate.

File: ruby/ql/lib/codeql/ruby/printAst.qll (hunk 2 above)
[the remark this answers is not in the document]
Comment (aibaars) [reply ruby/ql/lib/codeql/ruby/printAst.qll:119]: Good point, done!

**Review by hvitved — APPROVED** [review]

**Pull request merged**